### PR TITLE
feat(serialization): Alternative serialization approach

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordonjsmith@gmail.com>

## Checklist:
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit message includes a "fixes" reference if appropriate.
  - [ ] The commit is signed.
- [ ] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

**Comment by GordonSmith** [comment]
For discussion - proposed replacement for PropertyExt

File: tests/test-util/src/serialize.spec.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        expect(instance2.backColor()).to.equal("red");
+        expect(instance2.backColor.exists()).to.equal(true);
+        expect(instance2.backColor.modified()).to.equal(false);
+    });
+
+    it("functor value", function () {
+        const instance = new Derived2();
+
+        //  Note "w" is typed correctly  ---
+        instance.backColor(w => w.foreColor());
```
Comment (GordonSmith) [inline tests/test-util/src/serialize.spec.ts]: Note:  property values can now be functions...

File: packages/util/src/serialize.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    const metaProp = to_meta(prop);
+                    if ((this as any)[prop]) {
+                        (this as any)[prop] = (this as any)[prop].bind(this, __proto__[METAProp].func);
+                    } else {
+                        (this as any)[prop] = __proto__[METAProp].func.bind(this);
+                    }
+                    (this as any)[prop].reset = __proto__[METAProp].reset.bind(this, metaProp, METAProp);
+                    (this as any)[prop].modified = __proto__[METAProp].modified.bind(this, metaProp, METAProp);
+                    (this as any)[prop].exists = __proto__[METAProp].exists.bind(this, metaProp, METAProp);
+                });
```
Comment (GordonSmith) [inline packages/util/src/serialize.ts]: This loop is my main objection to this code - but I believe it is still better then the current implementations (I am trying to convince myself its ok...).

Comment (jbrundage) [reply packages/util/src/serialize.ts]: The use of lower and upper case "META" is confusing and makes this difficult to read.

File: packages/core/src/serialize.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                const retVal = {
+                    module: moduleName,
+                    class: className,
+                    properties: {} as any
+                };
+                for (const key in __proto__) {
+                    if (is_META(key)) {
+                        const property = from_META(key);
+                        if ((this as any).modified(property)) {
+                            retVal.properties[property] = (this as any)[property]();
```
Comment (jbrundage) [inline packages/core/src/serialize.ts]: Maybe this line should be:
```javascript
let val = (this as any)[property]();
if(val && val.serialize){
    val = val.serialize();
}
retVal.properties[property] = val;
```

File: packages/core/src/serialize.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                }
+                return retVal;
+            }
+
+            deserialize(_: SerializationObject) {
+                if (moduleName !== _.module || className !== _.class) throw new Error("Module / Class does not match");
+                for (const key in __proto__) {
+                    if (is_META(key)) {
+                        const property = from_META(key);
+                        (this as any)[property](_.properties[property]);
```
Comment (jbrundage) [inline packages/core/src/serialize.ts]: This is goes along with my suggestion for `serialize`. This line should become something like:
```javascript
let val = _.properties[property];
if(val && val.module){
    val = new Widget().deserialize(val);
}
(this as any)[property](val);
```

**Pull request closed without merging**